odoo/models.py

```diff
@@ -5533,16 +5578,18 @@ def diff(self, other):
                 if field_onchange.get(name):
                     record._onchange_eval(name, field_onchange[name], result)
 
-                # make a snapshot (this forces evaluation of computed fields)
-                snapshot1 = Snapshot(record, nametree)
+        # make a snapshot (this forces evaluation of computed fields)
+        snapshot1 = Snapshot(record, nametree)
 
-                # determine which fields have been modified
-                for name in nametree:
-                    if snapshot1[name] != snapshot0[name]:
-                        todo.append(name)
+        # determine which fields have been modified
+        for name in nametree:
+            if snapshot1[name] != snapshot0[name]:
+                todo.append(name)
 
         # determine values that have changed by comparing snapshots
-        self.invalidate_cache()
+
+        # FP TODO, do we need this?
+        # self.invalidate_cache()
```

Comment:
Yes we do need something like this.
You want to know which records have been deleted from database.
You have to invalidate their cache, because they don't have a value anymore.
Because of the `ondelete='cascade'`, other records have been deleted, but you don't know which ones (PostgreSQL does not tell you.)

Reply:
odoo can calculated the cascade tree (we know witch field is cascade or not), no ?